sympy/polys/rings.py

```diff
@@ -3005,3 +3179,295 @@ def sqf_list(f, all=False):
+            all_polys.extend(coeffs_i)
+
+            # Quick exit:
+            if any(len(c) == 1 for c in coeffs_i):
+                ring = polynomials[0].ring
+                domain = ring.domain
+                gcd = gcd_terms((all_polys + polynomials[i+1:]), ring, domain)
+                return [gcd], None
+
+def gcd_prs(polynomials):
+    """
+    Returns the greatest common divisor (GCD) of a set of polynomials.
+
+    Examples
+    ========
+
+    >>> from sympy.polys.rings import gcd_prs
+    >>> from sympy import ZZ, ring
+    >>> R, x, y = ring("x, y", ZZ)
+
+    >>> polynomials = x - y, x - y
+    >>> gcd_prs(polynomials)
+    x - y
+
+    """
+    ring = polynomials[0].ring
+    domain = ring.domain
+
+    if any(len(pi) == 1 for pi in polynomials):
+        return gcd_terms(polynomials, ring, domain)
+
+    polynomials, monom_gcd = monomial_extract(polynomials)
+
+    polynomials, common_symbols = _gcd_preprocess_polys(polynomials)
+
+    gcd = polynomials[0]
+    for pi in polynomials[1:]:
+        gcd = _gcd_prs(gcd, pi)
+        if gcd == domain.one:
+            break
+
+    if monom_gcd is not None:
+        gcd = gcd * monom_gcd
+
+    return gcd
+
+def _gcd_prs(p1, p2):
+    """
+    Returns the greatest common divisor (GCD) of two polynomials using the
+    Polynomial Resultant Sequences (PRS) method.
+
+    Examples
+    ========
+
+    >>> from sympy.polys.rings import _gcd_prs
+    >>> from sympy import ZZ, ring
+    >>> R, x, y = ring("x, y", ZZ)
+
+    >>> f = 4*x**2 + 8*x + 10
+    >>> g = 2*x**2 + 6*x + 10
+    >>> _gcd_prs(f, g)
+    2
+
+    """
+    x = p1.main_variable()
+
+    c1, pp1 = cont_prim(p1, x)
+    c2, pp2 = cont_prim(p2, x)
+
+    h = pp1.subresultants(pp2, x)[-1]
+    c = gcd_prs([c1, c2])
+
+    ring = p1.ring
+    domain = ring.to_domain()
+
+    _, h = cont_prim(h, x)
+
+    if not domain.is_Field:
+        return h*c
+    else:
+        return -h*c
+
+def _gcd_(f, g):
+    """Helper function for ``_gcd_prs``."""
+
+    ring = f.ring
+    domain = ring.domain
+
+    if not domain.is_Exact:
+        try:
+            exact = domain.get_exact()
+        except DomainError:
+            return domain.one, f, g
+
+        f = ring.dmp_convert(f, domain, exact)
+
+        g = ring.dmp_convert(g, domain, exact)
+
+        h, cff, cfg  = f._gcd_(g)
+
+        h = ring.dmp_convert(h, exact, domain)
+        cff = ring.dmp_convert(cff, exact, domain)
+        cfg = ring.dmp_convert(cfg, exact, domain)
+
+        return h, cff, cfg
+    elif domain.is_Field:
+        if domain.is_QQ and query('USE_HEU_GCD'):
+            try:
+                return ring.dmp_qq_heu_gcd(f, g)
+            except HeuristicGCDFailed:
+                pass
+
+        return f._gcd_I(g)
+    else:
+        if domain.is_ZZ and query('USE_HEU_GCD'):
+            try:
+                return ring.dmp_zz_heu_gcd(f, g)
+            except HeuristicGCDFailed:
+                pass
+
+        return f._gcd_I(g)
```

Comment:
Is it the correct way?

Reply:
It shouldn't be using any of the dense `dmp_*` functions.

Reply:
We don't literally need to copy this function from the dense implementation. What is needed is that the algorithmic steps that happen in the dense implementation should also happen in the sparse implementation so the way that RR is handled should be the same etc. If that was done correctly then there would be no test failures and no need to change any tests.

Reply:
Should I also create the `dmp_convert` function? I couldn't locate a function that serves a similar purpose.